lightning/src/ln/channelmanager.rs

```diff
@@ -7906,6 +7704,346 @@ where
+			payment_data: data,
+			payment_metadata,
+			incoming_cltv_expiry: outgoing_cltv_value,
+			phantom_shared_secret,
+			custom_tlvs,
+		}
+	} else {
+		return Err(InboundOnionErr {
+			err_code: 0x4000|0x2000|3,
+			err_data: Vec::new(),
+			msg: "We require payment_secrets",
+		});
+	};
+	Ok(PendingHTLCInfo {
+		routing,
+		payment_hash,
+		incoming_shared_secret: shared_secret,
+		incoming_amt_msat: Some(amt_msat),
+		outgoing_amt_msat: onion_amt_msat,
+		outgoing_cltv_value,
+		skimmed_fee_msat: counterparty_skimmed_fee_msat,
+	})
+}
+
+/// Peel one layer off an incoming onion, returning PendingHTLCInfo (either Forward or Receive).
+/// This does all the relevant context-free checks that LDK requires for payment relay or
+/// acceptance. If the payment is to be received, and the amount matches the expected amount for
+/// a given invoice, this indicates the UpdateAddHTLC, once fully committed in the channel, will
+/// generate an Event::PaymentReceived.
+pub fn peel_payment_onion<
+	NS: Deref,
+	L: Deref,
+	T: secp256k1::Verification,
+>(
+	msg: &msgs::UpdateAddHTLC, node_signer: &NS, logger: &L, secp_ctx: &Secp256k1<T>,
+	cur_height: u32, accept_mpp_keysend: bool,
+) -> Result<PendingHTLCInfo, InboundOnionErr>
+where
+	NS::Target: NodeSigner,
+	L::Target: Logger,
+{
+	let (hop, shared_secret, next_packet_details_opt) = 
+		decode_incoming_update_add_htlc_onion(msg, node_signer, logger, secp_ctx
+	).map_err(|_| InboundOnionErr {
+		msg: "Failed to decode update add htlc onion",
+		err_code: 0x4000 | 22,
+		err_data: Vec::new(),
+	})?;
+	Ok(match hop {
+		onion_utils::Hop::Forward{ next_hop_data, next_hop_hmac, new_packet_bytes } => {
+			if next_packet_details_opt.is_none() {
+				// Forward should always include the next hop details
+				return Err(InboundOnionErr {
+					msg: "Failed to decode update add htlc onion",
+					err_code: 0x4000 | 22,
+					err_data: Vec::new(),
+				});
+			};
+			let (next_packet_pubkey, _, _, outgoing_cltv_value) = next_packet_details_opt.unwrap();
+
+			if let Err((err_msg, code, _)) = check_incoming_htlc_cltv(
+				cur_height, outgoing_cltv_value, msg.cltv_expiry
+			) {
+				return Err(InboundOnionErr {
+					msg: err_msg,
+					err_code: code,
+					err_data: Vec::new(),
+				});
+			}
+			create_fwd_pending_htlc_info(
+				msg, next_hop_data, next_hop_hmac, new_packet_bytes, shared_secret,
+				Some(next_packet_pubkey)
+			)?
+		},
+		onion_utils::Hop::Receive(received_data) => {
+			create_recv_pending_htlc_info(
+				received_data, shared_secret, msg.payment_hash, msg.amount_msat, msg.cltv_expiry,
+				None, false, msg.skimmed_fee_msat, cur_height, accept_mpp_keysend,
+			)?
+		}
+	})
+}
+
+fn decode_incoming_update_add_htlc_onion<
+	NS: Deref,
+	L: Deref,
+	T: secp256k1::Verification,
+>(
+	msg: &msgs::UpdateAddHTLC, node_signer: &NS, logger: &L, secp_ctx: &Secp256k1<T>,
+) -> Result<(
+	onion_utils::Hop, [u8; 32], Option<(Result<PublicKey, secp256k1::Error>, u64, u64, u32)>
```

Comment:
`next_packet_details` should really be made into a struct rather than a long optional tuple. Can be saved for follow-up though.

Reply:
I went ahead and did it! Simple change